### ISSUE TITLE
The problem with pyserial reading from serial port

### ISSUE THREAD

**Issue by GreenToxicPain** [pyserial/pyserial#511]
Hello. I used to use pyserial for several projects (communicating with stm32, arduino or even virtual device via serial port). I tested python scripts on Linux, Windows (even on Linux on raspberry pi). Every time despite pyserial version I got some issues related to serial reading. The problem occurs if you try to read some fixed number of bytes using commands 'read(...)' or 'read_until(...)'. With some specific byte sequences to read the problem occurs as reading unsufficient bytes (e.g. you try to read [1, 2, 3] but got only [1, 2] and after reading ser.inWaiting() == 1) or serial port freezing. I solved this problem using code below:

class MySerial():
    def __init__(self, baudrate, port_name):
        self.read_buffer = []

        self.port = MySerial(
                port=port_name,
                baudrate=baudrate,
                parity=serial.PARITY_ODD,
                stopbits=serial.STOPBITS_TWO,
                bytesize=serial.SEVENBITS
            )

    def read(self, message_size):
        while len(self.read_buffer) < message_size:
            n = self.port.inWaiting()
            if n:
                self.read_buffer += self.port.read(n)
        data = self.read_buffer[:message_size]
        self.read_buffer = self.read_buffer[message_size:]
        return data

        # return self.port.read_until(size=message_size)

    def write(self, message):
        self.port.write(message)

In this code we check number of bytes to read (inWaiting) and read all of them.

**Comment by mrbell321** [comment]
Just browsing and... MySerial.__init__() creates an instance of MySerial?

**Comment by shariqkhan29** [comment]
while len(self.read_buffer) **<** message_size:
may be **<=** so it reads the whole message

**Comment by GreenToxicPain** [comment]
> while len(self.read_buffer) **<** message_size:
> may be **<=** so it reads the whole message

Nope, in that case we don't need to reed anything from serial port and can just return required bytes from the buffer
